tests/combine_span_test.cpp

```diff
@@ -0,0 +1,280 @@
+
+  return span;
+}
+
+/**  Mocked RNG that returns a set value deterministically */
+struct mock_rng {
+  double value;
+  double uniform_real_01() { return value; }
+};
+
+// helpers to use GoogleTest's parameterization system
+using Forwards =
+    std::integral_constant<nuts::Direction, nuts::Direction::Forward>;
+using Backwards =
+    std::integral_constant<nuts::Direction, nuts::Direction::Backward>;
+using Metropolis =
+    std::integral_constant<nuts::Update, nuts::Update::Metropolis>;
+using Barker = std::integral_constant<nuts::Update, nuts::Update::Barker>;
+
+using ForwardsMetropolis = std::tuple<Forwards, Metropolis>;
+using BackwardsMetropolis = std::tuple<Backwards, Metropolis>;
+using ForwardsBarker = std::tuple<Forwards, Barker>;
+using BackwardsBarker = std::tuple<Backwards, Barker>;
+
+}  // namespace span_test
+
+// macros to compare parts of Spans
+#define EXPECT_FORWARD_ENDPOINT_EQUAL(span_recieved, span_expected)      \
+  EXPECT_EQ(span_recieved.theta_fw_, span_expected.theta_fw_);           \
+  EXPECT_EQ(span_recieved.rho_fw_, span_expected.rho_fw_);               \
+  EXPECT_EQ(span_recieved.grad_theta_fw_, span_expected.grad_theta_fw_); \
+  EXPECT_EQ(span_recieved.logp_fw_, span_expected.logp_fw_);
+
+#define EXPECT_BACKWARD_ENDPOINT_EQUAL(span_recieved, span_expected)     \
+  EXPECT_EQ(span_recieved.theta_bk_, span_expected.theta_bk_);           \
+  EXPECT_EQ(span_recieved.rho_bk_, span_expected.rho_bk_);               \
+  EXPECT_EQ(span_recieved.grad_theta_bk_, span_expected.grad_theta_bk_); \
+  EXPECT_EQ(span_recieved.logp_bk_, span_expected.logp_bk_);
+
+#define EXPECT_SELECTED_POINT_EQUAL(span_recieved, input, new_logp) \
+  EXPECT_EQ(span_recieved.theta_select_, input.theta_select_);      \
+  EXPECT_EQ(span_recieved.grad_select_, input.grad_select_);        \
+  EXPECT_FLOAT_EQ(span_recieved.logp_, new_logp);
+
+// The forwards/backwards pieces of the new span are update agnostic
+template <typename T>
+class CombineSpansUpdateAgnostic : public ::testing::Test {};
+
+using BothUpdates = ::testing::Types<span_test::Metropolis, span_test::Barker>;
+TYPED_TEST_SUITE(CombineSpansUpdateAgnostic, BothUpdates);
+
+TYPED_TEST(CombineSpansUpdateAgnostic, ForwardEndpointsCorrect) {
+  constexpr auto update = TypeParam::value;
+
+  span_test::mock_rng rng{0};
+  nuts::SpanW<double> span_from = span_test::dummy_span(0.3);
+  nuts::SpanW<double> span_to = span_test::dummy_span(0.1);
+
+  nuts::SpanW<double> span_old = span_from;
+  nuts::SpanW<double> span_new = span_to;
+  nuts::SpanW<double> combined =
+      nuts::combine<update, nuts::Direction::Forward>(rng, std::move(span_old),
+                                                      std::move(span_new));
+
+  EXPECT_BACKWARD_ENDPOINT_EQUAL(combined, span_from);
+  EXPECT_FORWARD_ENDPOINT_EQUAL(combined, span_to);
+}
+
+TYPED_TEST(CombineSpansUpdateAgnostic, BackwardsEndpointsCorrect) {
+  constexpr auto update = TypeParam::value;
```

Comment:
This variable is never used unless it's some kind of indirection through `TYPED_TEST`.  I really find macros opaque in C++.  And I can't figure out what `TypeParam` is and thus can't figure out what `value` is.